Lib/unittest/mock.py

```diff
@@ -2666,15 +2659,18 @@ def create_autospec(spec, spec_set=False, instance=False, _parent=None,
         # descriptors don't have a spec
         # because we don't know what type they return
         _kwargs = {}
-    elif is_async_func:
-        if instance:
+    elif is_async:
+        if instance and isinstance(spec, FunctionTypes):
             raise RuntimeError("Instance can not be True when create_autospec "
                                "is mocking an async function")
         Klass = AsyncMock
     elif not _callable(spec):
         Klass = NonCallableMagicMock
-    elif is_type and instance and not _instance_callable(spec):
-        Klass = NonCallableMagicMock
+    elif is_type and instance:
+        if _is_async_obj(spec.__new__(spec)):
```

Comment:
Actually think this is a bad idea, especially on a relatively frequent codepath. Going to try to reimplement this without executing `__new__`.

Reply:
Fixed by adapting the isawaitable implementation for types.